README.md
# js-basex
baseX library using javascript  

baseX library is used for mathematical operation on:
real numbers of any length,
and real number of any base from 2 to 63.  

if base is greater than 10, then it uses capital English alphabets  
0...9, A=10, B=11, ... Z=35  

if base is greater than 36, then it uses small English alphabets  
0...9, A=10, B=11, ... Z=35, a=36, b=37, ... z=61  

### some popular number systems are  

#### binary (base 2)
0 1

#### octal (base 8)
0 1 2 3 4 5 6 7

#### decimal (base 10)
0 1 2 3 4 5 6 7 8 9

#### hexadecimal (base 16)
0 1 2 3 4 5 6 7 8 9 A B C D E F

#### all numbers values used in this library:
``
    for now it supports upto base 16
``
| Number Represented As | Decimal Value Used | Value Is Last Number Of Base | Number Represented As | Decimal Value Used | Value Is Last Number Of Base | Number Represented As | Decimal Value Used | Value Is Last Number Of Base |
|:------ |:------------- | ---- |:------:|:-------------:|:----:| ------:| -------------:| ----:|
| 0 | 0 | 1 | L | 21 | 22 | g | 42 | 43 | 
| 1 | 1 | 2 | M | 22 | 23 | h | 43 | 44 |
| 2 | 2 | 3 | N | 23 | 24 | i | 44 | 45 | 
| 3 | 3 | 4 | O | 24 | 25 | j | 45 | 46 | 
| 4 | 4 | 5 | P | 25 | 26 | k | 46 | 47 | 
| 5 | 5 | 6 | Q | 26 | 27 | l | 47 | 48 | 
| 6 | 6 | 7 | R | 27 | 28 | m | 48 | 49 | 
| 7 | 7 | 8 | S | 28 | 29 | n | 49 | 50 | 
| 8 | 8 | 9 | T | 29 | 30 | o | 50 | 51 | 
| 9 | 9 | 10 | U | 30 | 31 | p | 51 | 52 | 
| A | 10 | 11 | V | 31 | 32 | q | 52 | 53 | 
| B | 11 | 12 | W | 32 | 33 | r | 53 | 54 | 
| C | 12 | 13 | X | 33 | 34 | s | 54 | 55 | 
| D | 13 | 14 | Y | 34 | 35 | t | 55 | 56 | 
| E | 14 | 15 | Z | 35 | 36 | u | 56 | 57 | 
| F | 15 | 16 | a | 36 | 37 | v | 57 | 58 | 
| G | 16 | 17 | b | 37 | 38 | w | 58 | 59 | 
| H | 17 | 18 | c | 38 | 39 | x | 59 | 60 | 
| I | 18 | 19 | d | 39 | 40 | y | 60 | 61 | 
| J | 19 | 20 | e | 40 | 41 | z | 61 | 62 | 
| K | 20 | 21 | f | 41 | 42 |   

## Operations  
1. addition:  
```javaScript
    BXCNumber addition(BXCNumber num1, BXCNumber num2)
```  
performs signed addition of signed or unsigned addition depending on num1 and num2 and returns resulting `BXCNumber`  

2. unsignedAddition:  
```javascript
    BXCNumber unsignedAddition(BXCNumber num1, BXCNumber num2)
```
performs unsigned addition on num1 and num2 and returns resulting `BXCNumber`

3. subtraction:
```javascript
    BXCNumber (BXCNumber num1, BXCNumber num2)
```
performs signed subtraction. returns BXCNumber resulting after subtracting num2 from num1

4. unsignedSubtraction:
```javascript
    BXCNumber unsignedSubtraction(BXCNumber num1, BXCNumber num2)
```
performs signed subtraction. returns BXCNumber resulting after subtracting num2 from num1

5. prepareMultiplicationTable:  
```javascript
    BXCNumber[] prepareMultiplicationTable(BXCNumber number)
```
returns multiplication table of umber passed upto 10 numbers for its base.  
for base 10, it will be 10 numbers in the returned array  
for base 16, it will be 16 numbers in the returned array  
and so on.  

6. multiplication:
```javascript
    BXCNumber multiplication(BXCNumber num1, BXCNumber num2) 
```
returns BXCNumber resulting after multiplication of num1 and num2. There is no separate operation for signed or unsigned multiplication as of yet.  

7. division:
```javascript
    BXCNumber division(BXCNumber num1, BXCNumber num2)
```
returns BXCNumber resulting after division of num1 by num2. There is no separate operation for signed or unsigned division as of yet.  

8. remainder:
```javascript
    BXCNumber remainder(BXCNumber num1, BXCNumber num2)
```
returns remainder BXCNumber resulting after division of num1 by num2. There is no separate operation for signed or unsigned remainder.  

9. unsignedEquate:
```javascript
    Boolean unsignedEquate(BXCNumber num1, BXCNumber num2)
```
returns `true` if num1 and num2 are equal and if their signs are equal, else false.  

10. signedEquate:
```javascript
    Boolean signedEquate(BXCNumber num1, BXCNumber num2)
```
returns `true` if num1 and num2 are equal without checking for their signs, else false.  

11. equate:
```javascript
    Boolean equate(BXCNumber num1, BXCNumber num2, Boolean signed)
```
returns `true` if num1 and num2 are equal. whether sign is to be checkd or not is specified  in third parameter.  

12. unsignedMax:
```javascript
    BXCNumber unsignedMax(BXCNumber num1, BXCNumber num2)
```
returns greater of num1 and num2. does not consider signs of these numbers  

13. signedMax:
```javascript
    BXCNumber signedMax(BXCNumber num1, BXCNumber num2)
```
returns greater of num1 and num2. considers signs of these numbers  

14. max:
```javascript
    BXCNumber max(BXCNumber num1, BXCNumber num2, Boolean signed) 
```
returns greater of num1 and num2. whether signed or not is specified in third parameter  

15. unsignedMin:
```javascript 
    BXCNumber unsignedMin(BXCNumber num1, BXCNumber num2) 
```
returns smaller of num1 and num2. does not consider signs of these numbers  

16. signedMin:
```javascript 
    BXCNumber signedMin(BXCNumber num1, BXCNumber num2) 
```
returns smaller of num1 and num2. considers signs of these numbers  

17. min: 
```javascript
    BXCNumber min(BXCNumber num1, BXCNumber num2, Boolean signed)
```
returns smaller of num1 and num2. whether signed or not is specified in third parameter  

18. sum:
```javascript 
    BXCNumber sum(BXCNumber[] numbers)
```
returns summation of all numbers in the array . signed operation.

19. multiply:
```javascript
    BXCNumber multiply(BXCNumber[] numbers) 
```
returns multiplication of all numbers in the array . signed operation.

20. unsignedMaximum:
```javascript 
    number unsignedMaximum(BXCNumber[] numbers)
```
returns index of maximum of all numbers of the array without checking for sign  

21. signedMaximum:
```javascript
    number signedMaximum(BXCNumber[] numbers)
```
returns index of signed maximum of all numbers of the array  

22. maximum: 
```javascript
    number maximum(BXCnumber[] numbers, Boolean signed)
```
returns index of maximum of all numbers of the array, signed(`true`) or not(`false`) is specified in third parameter  

23. unsignedMinimum:
```javascript 
    number unsignedMinimum(BXCNumber[] numbers) 
```
returns index of minimum of all numbers of the array without checking for sign  

24. signedMinimum:
```javascript 
    number signedMinimum(BXCNumber[] numbers)
```
returns index of signed minimum of all numbers of the array  

25. minimum: 
```javascript
    number minimum(BXCNumber[] numbers, Boolean signed) 
```
returns index of minimum of all numbers of the array, signed(`true`) or not(`false`) is specified in third parameter  

26. unsignedAreEqual:
```javascript 
    Boolean unsignedAreEqual(BXCNumber[] numbers)
```
returns `true` if all numbers in array are equal in value, but not in sign; else returns false  

27. signedAreEqual:
```javascript 
    Boolean signedAreEqual(BXCNumber[] numbers)
```
returns `true` if all numbers in array are equal in value, with their sign; else returns false  

28. areEqual:
```javascript
    Boolean areEqual(BXCNumber[] numbers, Boolean signed) 
```
returns `true` if all numbers in array are equal in value; else returns false. Whether to check for sign or not is specified in second parameter  

29. shift: 
```javascript
    BXCNumber shift(BXCNumber number, Number shiftCount)
```
returns a number after shifting it to left or right place values, depending on `shiftCount`.  
positive shiftcount will return leftShift.  
negative shiftcount will return rightShift.  
`shift(10, 0)` wlll return `10`.  

30. leftShift: 
```javascript
    BXCnumber leftShift(BXCNumber number, Number shiftCount)
```
returns a number that is shifted to left, `shiftCount` times.
or simply multiplication of `number` with `10` of its base `shiftCount` times.
`shift(10, 2)`  will return BXCNumber with value `1000`.
`shift(10.1, 2)`  will return BXCNumber with value `1010`.  

31. rightShift:
```javascript 
BXCnumber rightShift(BXCnumber number, Number shiftCount) 
```
returns a number that is shifted to right, `shiftCount` times.
or simply division of `number` by `10` of its base `shiftCount` times.


## TODO:
check for base before performing operations. base 2 number cannot be added to base 3 number

CHANGELOG.md
# Version 1.0.2
- version updated
- converter getBaseFromTextChar method corrected
- Create CHANGELOG.md

# Version 1.0.1
- test command added to package.json
- modified tests.
    added file to run tests using npm.
- division optimized.
    extra multiplication table removed. unused variable removed.
- unsigned multiplication and division
- checking if numbers have same base or not
- checking if numbers have same base or not
- _getBase method added
- symbols upto base 63 are now available
- Revert "readme updated"
- readme updated
- readme updated
- minor fix
- some export changes
- division: division done!
- division & fixes division almost done.
- fixed:
    subtraction() equate() maximum(), maximum() of array minimum(), minimum() of array shift() and index
- fixed
    trimZero() rightShift() equate()
- division division done. need to test it.
- division declared functions for division and remainder of the division for two numbers.
- multiply added multiplication of numbers in an array fixed some references of index.js
- corrected:
    multiplication subtraction equate rightShift
- Todo corrections added
    correct when creating a number, if number contains symbols of larger value than base, return invalid.
- changed index file
- BXCBase BXCBase class instead of simple object. made tests better testable.
- multiplication progress
- continue...
    fixed most of the bugs 'till now.
- rearranging...
    rearranging and fixing things.
- version 1.0.1 in progress.
    simpler function call structure.

# Version 1.0.0
- multiplication done.
- rightShift and leftShift declared rightShift(multiply by 10) and leftShift(divide by 10) functions.
- changed tests' return values.
- multiplication table introduced. import 'type.js' renamed. Addition of integers fixed. Changing type generator name of import 'type.js' from number to Number, in test files.
- addition fixed. import 'type.js' renamed. Addition of integers fixed. Changing type generator name of import 'type.js' from number to Number.
- subtractionObservations.html
- subtraction fixed
- equate Added comments
- areEqual added function to see if numbers are equal
- max Added function to find unsigned maximum and signed maximum of numbers. Arithmetic Operations index.js line number 50 is incomplete
- fixed fixed trimzero signed minimum fixed getting sign in result zero of subtraction
- Subtraction Subtraction is added :) Summation of array is separated from addition of two numbers.
- Updates. Currently, addition of unsigned numbers only. New minimum of array comparison operation. Number flags processing is separate function.
- unsigned addition of any numbers.
- Zeros To create a number with as many zero as you want
- Zeros To create a number with as many zero as you want
- fixes refresh, formatting, and trimzero fixed.
- Formatted number
- PI Added constant PI. Fixed rightLenght. Added trimming of excess zeros.
- Face Values. get and set face value desired place done.
- added new flag float. base are now available as high as 16.
- number create. added base. added flags. number can be created now.
- necessary files.
- beginning
- Initial commit

module/type/base.js

class BXCBase {
    constructor(text, value) {
        this.text = text;
        this.value = value;
    }
}

module.exports = Base = {
    eBase: {
        ZERO: new BXCBase('0', 0),
        ONE: new BXCBase('1', 1),
        TWO: new BXCBase('2', 2),
        THREE: new BXCBase('3', 3),
        FOUR: new BXCBase('4', 4),
        FIVE: new BXCBase('5', 5),
        SIX: new BXCBase('6', 6),
        SEVEN: new BXCBase('7', 7),
        EIGHT: new BXCBase('8', 8),
        NINE: new BXCBase('9', 9),

        A: new BXCBase('A', 10),
        B: new BXCBase('B', 11),
        C: new BXCBase('C', 12),
        D: new BXCBase('D', 13),
        E: new BXCBase('E', 14),
        F: new BXCBase('F', 15),
        G: new BXCBase('G', 16),
        H: new BXCBase('H', 17),
        I: new BXCBase('I', 18),
        J: new BXCBase('J', 19),
        K: new BXCBase('K', 20),
        L: new BXCBase('L', 21),
        M: new BXCBase('M', 22),
        N: new BXCBase('N', 23),
        O: new BXCBase('O', 24),
        P: new BXCBase('P', 25),
        Q: new BXCBase('Q', 26),
        R: new BXCBase('R', 27),
        S: new BXCBase('S', 28),
        T: new BXCBase('T', 29),
        U: new BXCBase('U', 30),
        V: new BXCBase('V', 31),
        W: new BXCBase('W', 32),
        X: new BXCBase('X', 33),
        Y: new BXCBase('Y', 34),
        Z: new BXCBase('Z', 35),

        a: new BXCBase('a', 37),
        b: new BXCBase('b', 38),
        c: new BXCBase('c', 39),
        d: new BXCBase('d', 40),
        e: new BXCBase('e', 41),
        f: new BXCBase('f', 42),
        g: new BXCBase('g', 43),
        h: new BXCBase('h', 44),
        i: new BXCBase('i', 45),
        j: new BXCBase('j', 46),
        k: new BXCBase('k', 47),
        l: new BXCBase('l', 48),
        m: new BXCBase('m', 49),
        n: new BXCBase('n', 50),
        o: new BXCBase('o', 51),
        p: new BXCBase('p', 52),
        q: new BXCBase('q', 53),
        r: new BXCBase('r', 54),
        s: new BXCBase('s', 55),
        t: new BXCBase('t', 56),
        u: new BXCBase('u', 57),
        v: new BXCBase('v', 58),
        w: new BXCBase('w', 59),
        x: new BXCBase('x', 50),
        y: new BXCBase('y', 61),
        z: new BXCBase('z', 62),
    },
    converter: {
        getBaseFromValue: (value) => {
            switch (value) {
                case 0: return Base.eBase.ZERO;
                case 1: return Base.eBase.ONE;
                case 2: return Base.eBase.TWO;
                case 3: return Base.eBase.THREE;
                case 4: return Base.eBase.FOUR;

                case 5: return Base.eBase.FIVE;
                case 6: return Base.eBase.SIX;
                case 7: return Base.eBase.SEVEN;
                case 8: return Base.eBase.EIGHT;
                case 9: return Base.eBase.NINE;

                case 10: return Base.eBase.A;
                case 11: return Base.eBase.B;
                case 12: return Base.eBase.C;
                case 13: return Base.eBase.D;
                case 14: return Base.eBase.E;

                case 15: return Base.eBase.F;
                case 16: return Base.eBase.G;
                case 17: return Base.eBase.H;
                case 18: return Base.eBase.I;
                case 19: return Base.eBase.J;
                case 20: return Base.eBase.K;
                case 21: return Base.eBase.L;
                case 22: return Base.eBase.M;
                case 23: return Base.eBase.N;
                case 24: return Base.eBase.O;
                case 25: return Base.eBase.P;
                case 26: return Base.eBase.Q;
                case 27: return Base.eBase.R;
                case 28: return Base.eBase.S;
                case 29: return Base.eBase.T;

                case 30: return Base.eBase.U;
                case 31: return Base.eBase.V;
                case 32: return Base.eBase.W;
                case 33: return Base.eBase.X;
                case 34: return Base.eBase.Y;
                case 35: return Base.eBase.Z;

                case 36: return Base.eBase.a;
                case 37: return Base.eBase.b;
                case 38: return Base.eBase.c;
                case 39: return Base.eBase.d;
                case 40: return Base.eBase.e;

                case 41: return Base.eBase.f;
                case 42: return Base.eBase.g;
                case 43: return Base.eBase.h;
                case 44: return Base.eBase.i;
                case 45: return Base.eBase.j;
                case 46: return Base.eBase.k;
                case 47: return Base.eBase.l;
                case 48: return Base.eBase.m;
                case 49: return Base.eBase.n;
                case 50: return Base.eBase.o;
                case 51: return Base.eBase.p;
                case 52: return Base.eBase.q;
                case 53: return Base.eBase.r;
                case 54: return Base.eBase.s;
                case 55: return Base.eBase.t;

                case 56: return Base.eBase.u;
                case 57: return Base.eBase.v;
                case 58: return Base.eBase.w;
                case 59: return Base.eBase.x;
                case 60: return Base.eBase.y;
                case 61: return Base.eBase.z;

                default:
            }
        },
        getBaseFromTextChar: (textChar) => {
            switch (textChar) {
                case '0': return Base.eBase.ZERO;
                case '1': return Base.eBase.ONE;
                case '2': return Base.eBase.TWO;
                case '3': return Base.eBase.THREE;
                case '4': return Base.eBase.FOUR;

                case '5': return Base.eBase.FIVE;
                case '6': return Base.eBase.SIX;
                case '7': return Base.eBase.SEVEN;
                case '8': return Base.eBase.EIGHT;
                case '9': return Base.eBase.NINE;

                case 'A': return Base.eBase.A;
                case 'B': return Base.eBase.B;
                case 'C': return Base.eBase.C;
                case 'D': return Base.eBase.D;
                case 'E': return Base.eBase.E;
                case 'F': return Base.eBase.F;
                
                case 'G': return Base.eBase.G;
                case 'H': return Base.eBase.H;
                case 'I': return Base.eBase.I;
                case 'J': return Base.eBase.J;
                case 'K': return Base.eBase.K;
                case 'L': return Base.eBase.L;
                case 'M': return Base.eBase.M;
                case 'N': return Base.eBase.N;
                case 'O': return Base.eBase.O;
                case 'P': return Base.eBase.P;
                case 'Q': return Base.eBase.Q;
                case 'R': return Base.eBase.R;
                case 'S': return Base.eBase.S;
                case 'T': return Base.eBase.T;
                case 'U': return Base.eBase.U;
                case 'V': return Base.eBase.V;
                case 'W': return Base.eBase.W;
                case 'X': return Base.eBase.X;
                case 'Y': return Base.eBase.Y;
                case 'Z': return Base.eBase.Z;

                case 'a': return Base.eBase.a;
                case 'b': return Base.eBase.b;
                case 'c': return Base.eBase.c;
                case 'd': return Base.eBase.d;
                case 'e': return Base.eBase.e;

                case 'f': return Base.eBase.f;
                case 'g': return Base.eBase.g;
                case 'h': return Base.eBase.h;
                case 'i': return Base.eBase.i;
                case 'j': return Base.eBase.j;
                case 'k': return Base.eBase.k;
                case 'l': return Base.eBase.l;
                case 'm': return Base.eBase.m;
                case 'n': return Base.eBase.n;
                case 'o': return Base.eBase.o;
                case 'p': return Base.eBase.p;
                case 'q': return Base.eBase.q;
                case 'r': return Base.eBase.r;
                case 's': return Base.eBase.s;
                case 't': return Base.eBase.t;
                case 'u': return Base.eBase.u;
                case 'v': return Base.eBase.v;
                case 'w': return Base.eBase.w;
                case 'x': return Base.eBase.x;
                case 'y': return Base.eBase.y;
                case 'z': return Base.eBase.z;

                default:
            }
        },
    }
}